.github/workflows/release.yml

```diff
@@ -0,0 +1,172 @@
+name: Publish to crates.io
+
+# on:
+#   workflow_dispatch:
+#     inputs:
+#       version:
+#         description: 'Version to publish (without v prefix, e.g. 1.2.3)'
+#         required: true
+#         type: string
+#       release_notes:
+#         description: 'Release notes (Markdown supported)'
+#         required: false
+#         type: string
+#         default: 'See repository for changes in this release.'
+#       dry_run:
+#         description: 'Dry run mode (create branch but skip tag creation and use cargo publish --dry-run)'
+#         required: false
+#         type: boolean
+#         default: false
+
+on:
+  pull_request: # Test trigger
+
+env:
+  RUSTFLAGS: -Dwarnings
+  RUST_BACKTRACE: 1
+  # Change to specific Rust release to pin
+  rust_stable: stable
+  rust_nightly: nightly-2024-07-07
+  # When updating this, also update relevant docs
+  rust_min: '1.81'
+  RELEASE_ROLE_ARN: ${{ secrets.RELEASE_ROLE_ARN }}
+  TOKEN_KEY_NAME: ${{ secrets.TOKEN_KEY_NAME }}
+  AWS_REGION: us-west-2
+  # Test values
+  TEST_VERSION: '1.0.0'
+  TEST_RELEASE_NOTES: 'Test release notes for version 1.0.0'
+  TEST_DRY_RUN: 'true'
+
+permissions:
+  contents: write  # Write permission to allow pushing to the repository
+  id-token: write  # This is required for requesting the JWT
+
+jobs:
+  publish:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Install Rust ${{ env.rust_nightly }}
+        uses: dtolnay/rust-toolchain@stable
+        with:
+            toolchain: ${{ env.rust_stable }}
+
+      - name: Install cargo-nextest
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-nextest
+
+      - uses: Swatinem/rust-cache@v2
+
+      - name: Display version to be published
+        run: |
+          echo "Publishing version ${{ env.TEST_VERSION }}"
+          echo "Dry run: ${{ env.TEST_DRY_RUN }}"
+
+      - name: Configure Git
+        run: |
+          git config --global user.name "GitHub Actions"
+          git config --global user.email "actions@github.com"
+
+      - name: Update version in Cargo.toml
```

Comment:
Do we want to be auto updating versions and tagging in the release? Given we aren't likely to be releasing on a frequent cadence (compared to SDKs for example) I feel like we could probably get away with:

1. Manually updating the version and creating the appropriate tag (with/without a script depending on how many steps there are) and changelog entries. 
2. Add a pre-flight release check in here that the version we are trying to release doesn't already exist

Reply:
I started with an action to run when the tag was cut, then I realize we need to update the version in code first, and make sure the version we cut in github matches the version we have in code, which is easy to get wrong, IMO.

Reply:
Then a release shell script. I think having the code already staged for release will simplify this workflow (IMHO anyway).

Reply:
Just want to double check I got you, you mean we manually run a release script to make the commit ready for release (changing the version and add changlog), then after we merge the commit, we cut a release tag to trigger the publish to crate.io?

It's some extra steps and manually things to make sure we have the right version and steps to follow, but we can start from there.

In aws-crt-cpp, we have some like https://github.com/awslabs/aws-crt-cpp/blob/main/update-version.sh
which runs after the release is cut, then update the code -> make a commit -> delete the old tag -> cut new tag point to the commit.

Reply:
Yeah was just trying to simplify a bit. I'll leave it up to you but in my head it seemed easier to:

1. Manually update the version, changelog via PR and then tag it
2. Trigger CI to release and specify the tag to release